---
layout: default
title: Single cell RNA-seq analysis course
---


# Single cell RNA-seq analysis workshop 

On-line course,  2021-01-25 - 2021-01-29 (9.00-16.00)

National workshop open for PhD students, postdocs, researchers and other employees within Swedish academia. This course is run by the National Bioinformatics Infrastructure Sweden (NBIS).

Due to the COVID-19 situation, the 2021 course will be held on-line. 


### Important dates

Application open: October 20, 2020

Application deadline: December 16, 2020

Confirmation to accepted students: December 22, 2020

Responsible teachers: Åsa Björklund and Paulo Czarnewski.

Please contact edu.sc@nbis.se for course specific questions.

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/410/list.svg" width="40" height="40"> Course content
***

The course will cover the basic steps in single cell RNAseq (scRNAseq) processing and data analysis, with lectures and practical exercises. Topics covered will include:

* An overview of the current scRNAseq technologies
* Basic overview of pipelines for processing raw reads into expression values
* Quality control of scRNAseq data
* Dimensionality reduction and clustering techniques
* Data normalization
* Differential gene expression for scRNAseq data
* Celltype prediction
* Trajectory analysis
* Comparison of different analysis pipelines such as Seurat, Scran and Scanpy


<br/>

##### <img border="0" src="https://www.svgrepo.com/show/3874/contact-form.svg" width="40" height="40"> Application
***

##### How to apply

Fill in this [application](https://forms.gle/dNZPunSwyZn2M9hG9) by December 16th.

##### Course fee

This online training event has no fee. However, if you accept a position at the workshop and do not participate (no-show) you will be invoiced 2000 SEK.

*Please note that NBIS cannot invoice individuals


##### Who can apply?

This is a national course. The course is open for PhD students, postdocs, group leaders and core facility staff within all Swedish universities. We do accept application from other countries, but give priority to applicants from Swedish universities prior to applicants from industry and academics from other countries.

Please note that NBIS training events do not provide any formal university credits. The training content is estimated to correspond to a certain number of credits, however the estimated credits are just guidelines. If formal credits are crucial, the student needs to confer with the home department before submitting a course application in order to establish whether the course is valid for formal credits or not.


##### Entry requirements

Practical exercises can be performed using R or Python, so we only accept students with previous experience in one of those programming languages.

Required for being able to follow the course and complete the practial exercises:

* Basic knowledge in R/Python and command line (bash).
* Be able to use your own computer with a web camera and R or Python installed for the practical computational exercises. Instructions on installation will be sent by email to accepted participants.
* Programming/scripting experience is required (in R or python).
* Basic understanding of NGS technologies and RNA-sequencing data.


* **Desirable**: Previous experience with RNA-seq analysis and/or participation in NGS/RNA-seq course is an advantage.

Due to limited space the course can accommodate maximum of 25 participants. If we receive more applications, participants will be selected based on several criteria. Selection criteria include correct entry requirements, motivation to attend the course as well as gender and geographical balance.

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/158264/schedule.svg" width="40" height="40"> Schedule
***

Course schedule can be found here: [schedule](schedule.md) (Will be updated, for now is last years schedule)

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/6672/exercise.svg" width="40" height="40"> Exercises
***

All exercises can be found at [Exercises](exercises.md). (TBA)

For working on Uppmax: to use the allocations we have for the course, please look [here](login.md).

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/19652/maths-class-materials-cross-of-a-pencil-and-a-ruler.svg" width="40" height="40"> Precourse material
***

Please read carefully the [Precourse material](precourse.md) before the course start.

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/38706/group-of-people.svg" width="40" height="40"> Course leaders
***

If you have any questions regarding the course, please email to edu.sc@nbis.se.

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/19262/link.svg" width="40" height="40"> Useful links
***

* [Single cell RNA-seq course at from Hemberg lab](https://scrnaseq-course.cog.sanger.ac.uk/website/index.html)
* [Single cell RNA-seq course in Python](https://chanzuckerberg.github.io/scRNA-python-workshop/intro/about)
* [Single cell RNA-seq course at Broad](https://broadinstitute.github.io/2019_scWorkshop/)
* [Repository listing many scRNA-seq tools](https://github.com/seandavi/awesome-single-cell)
* [SingleCellExperiment objects for many datasets](https://hemberg-lab.github.io/scRNA.seq.datasets/)
* [Conquer datasets - many different datasets based on a salmon pipeline](http://imlspenticton.uzh.ch:3838/conquer/)
* [The Human Cell Atlas project](https://www.humancellatlas.org/)
* [The github repository for this course. Also contains the R markdown scripts that was used to generate all the exercises](https://github.com/NBISweden/workshop-scRNAseq)
* [Bitbucket repository for QC report scripts](https://bitbucket.org/asbj/qc-summary_scrnaseq)
* [Bitbucket repository for the NBIS scRNAseq pipeline](https://bitbucket.org/scilifelab-lts/lts-workflows-sm-scrnaseq)

<br/>

##### <img border="0" src="https://www.svgrepo.com/show/83468/navigation-history-interface-symbol-of-a-clock-with-an-arrow.svg" width="40" height="40"> Past courses
***

* [Stockholm January 27-30, 2020](https://nbisweden.github.io/workshop-archive/workshop-scRNAseq/2020-01-27/)
* [Stockholm February 4-6, 2019](https://nbisweden.github.io/workshop-archive/workshop-scRNAseq/2019-02-04/)
* [Stockholm May 21-23, 2018](https://nbisweden.github.io/workshop-archive/workshop-scRNAseq/2018-05-21/)
* [Uppsala October 2-4, 2017](https://scilifelab.github.io/courses/scrnaseq/1710/)

<br/>

<div style="text-align: right; font-size: 5px"> Icons are provided from [www.svgrepo.com](www.svgrepo.com) </div>
